src/github.com/t3chguy/riot-static/riot-static.go

```diff
@@ -0,0 +1,234 @@
+}
+
+func GetPublicRoom(c *gin.Context) { // ~/room/:roomId/chat
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "room.html", gin.H{
+		"Room": data.Rooms[roomId],
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomServers(c *gin.Context) { // ~/room/:roomId/servers
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "room_servers.html", gin.H{
+		"Room": data.Rooms[roomId],
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMembers(c *gin.Context) { // ~/room/:roomId/members
+	roomId := c.Param("roomId")
+
+	//var page int
+	//var err error
+	//if page, err = strconv.Atoi(c.DefaultQuery("page", "1")); err != nil {
+	//	page = 1
+	//}
+
+	//pageSize := 20
+
+	data.RLock()
+	length := len(data.Rooms[roomId].MemberInfo)
+	//skip, end := paginate(page, pageSize, length)
+	c.HTML(http.StatusOK, "room_members.html", gin.H{
+		"Room": data.Rooms[roomId],
+		//"MemberInfo": data.Rooms[roomId].Members[skip:end],
+		"NumMembers": length,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomPowerLevels(c *gin.Context) { // ~/room/:roomId/power_levels
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "power_levels.html", gin.H{
+		"PowerLevels": data.Rooms[roomId].PowerLevels,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMember(c *gin.Context) { // ~/room/:roomId/members/:mxid
+	roomId := c.Param("roomId")
+	mxid := c.Param("mxid")
+
+	data.RLock()
+	if memberInfo := data.Rooms[roomId].MemberInfo[mxid]; memberInfo != nil {
+		c.HTML(http.StatusOK, "member_info.html", gin.H{
+			"RoomID":     roomId,
+			"MXID":       mxid,
+			"MemberInfo": memberInfo,
+		})
+	} else {
+		c.AbortWithStatus(http.StatusNotFound)
+	}
+	data.RUnlock()
+}
+
+var data = struct {
+	resync.Once
+	sync.RWMutex
+	NumRooms int
+	Ordered  []*Room
+	Rooms    map[string]*Room
+}{}
+
+func LoadPublicRooms() {
+	// @TODO: fix this.
+	data.Lock()
+	fmt.Println("Loading public publicRooms")
+	resp, err := cli.PublicRooms(0, "", "")
+
+	if err == nil {
+		b := []*Room{}
+		c := map[string]*Room{}
+
+		// filter on actually WorldReadable publicRooms
+		for _, x := range resp.Chunk {
+			if x.WorldReadable {
+				var room *Room
+				if data.Rooms[x.RoomId] != nil {
+					room = data.Rooms[x.RoomId]
+				} else {
+					room = NewRoom(x)
+				}
+				b = append(b, room)
+				c[x.RoomId] = room
+			}
+		}
+
+		//data.Lock()
+		data.Rooms = c
+		data.NumRooms = len(b)
+		// copy order so we don't encounter slice hell
+		data.Ordered = make([]*Room, data.NumRooms)
+		copy(data.Ordered, b)
+
+		data.Unlock()
+	}
+
+	if err != nil {
+		panic(err)
+	}
+}
+
+func main() {
+	setupCli()
+
+	router := gin.Default()
+	router.SetHTMLTemplate(tpl)
+	router.Static("/assets", "./assets")
+
+	router.Use(func(c *gin.Context) {
+		data.Once.Do(LoadPublicRooms)
+	})
+
+	router.GET("/", GetPublicRoomsList)
+
+	roomRouter := router.Group("/room/")
+	{
+		roomRouter.Use(func(c *gin.Context) {
+			roomId := c.Param("roomId")
+			if data.Rooms[roomId] == nil {
```

Comment:
You cannot do this. Any map reads/writes across goroutines must be protected with locks. It will work some of the time, but Go will panic if it can detect it.

> Maps are not safe for concurrent use: it's not defined what happens when you read and write to them simultaneously. If you need to read from and write to a map from concurrently executing goroutines, the accesses must be mediated by some kind of synchronization mechanism. 

https://blog.golang.org/go-maps-in-action